source/_drafts/Introducing-Thumbcoil.md

```diff
@@ -0,0 +1,77 @@
+
+Early on we managed to figure out how to coax FFmpeg into creating fragmented-MP4s from TS segments that would play back in a browser with media-source-extensions (at the time that meant only Chrome.) But we needed a simple way to compare the output of our transmuxer with what was produced by FFmpeg. The comparison had to be aware of the MP4 format since the two outputs are extremely unlikely to be byte-identical.
+
+{% pullquote left %}
+MP4 files are composed of **boxes** - hierarchical logical units that, conveniently, all start with a 32-bit length and a 32-bit box-type. Boxes will often contain other sub-boxes.
+{% endpullquote %}
+
+The answer to that problem was to build an "mp4-inspector" - a tool that would parse MP4 and display a sort of JSON-like dump of any relevant boxes and their contents. By generating a dump of the output from Mux.js and comparing it to a known-good fragment generated with FFmpeg, we could see where our transmuxer's output differed.
+
+The "mp4-inspector" was built as a web page so that we can have a graphical color-coded diff of the two segments. Over time the page gained an video element and we started appending the results of transmuxing segments directly into the video element's media-source to aid in instant feedback and validation of changes to Mux.js.
+
+## A Brave New World
+{% pullquote right %}
+A **media container** such as MP4 encapsulates the video and audio stream. It has metadata describing the streams, timing information for each frame, and the stream data itself.
+{% endpullquote %}
+
+As development continued, we would sometimes encounter streams that would fail in new and interesting ways. Some of these failures were, admittedly, due to bugs in Mux.js but as development continued and Mux.js itself became more robust they were increasingly caused by problems with the streams or issues with a particular implementation of the media-source-extensions specification.
+
+It eventually dawned on us that we really needed to learn more about what was happening inside of those videos. We needed to see not just what was happening at the media container level but we had to go deeper - we needed to peek into the video data itself. For that purpose we created Thumbcoil.
+
+{% pullquote left %}
+Inside of a container, video and audio are contained in data called **bitstreams**. Bitstreams are the data produced by encoders to represent the audio signals or video frames. Some common bitstreams are AAC for audio and H.264 for video.
+{% endpullquote %}
+
+## Thumbcoil
+
+Thumbcoil is a suite of tools designed to give you a peek into the internals of H.264 video bitstreams contained inside either an MPEG4 or MPEG2-TS container files. Using the tools in Thumbcoil you can get a detailed view of the internal structure of the two supported media container formats.
```

Comment:
Also, above you called them just MP4s and TS segments but here you're calling them MPEG4 and MPEG2-TS container files.